src/wp-includes/class-wp-theme-json-resolver.php

```diff
@@ -122,75 +140,243 @@ public static function get_core_data() {
+					'post_status'  => 'publish',
+					'post_title'   => __( 'Custom Styles', 'default' ),
+					'post_type'    => $post_type_filter,
+					'post_name'    => 'wp-global-styles-' . urlencode( wp_get_theme()->get_stylesheet() ),
+					'tax_input'    => array(
+						'wp_theme' => array( wp_get_theme()->get_stylesheet() ),
+					),
+				),
+				true
+			);
+			$user_cpt    = get_post( $cpt_post_id, ARRAY_A );
+		}
+
+		return $user_cpt;
+	}
+
+	/**
+	 * Returns the user's origin config.
+	 *
+	 * @since 5.9.0
+	 *
+	 * @return WP_Theme_JSON Entity that holds user data.
+	 */
+	public static function get_user_data() {
+		if ( null !== self::$user ) {
+			return self::$user;
+		}
+
+		$config   = array();
+		$user_cpt = self::get_user_data_from_custom_post_type();
+		if ( array_key_exists( 'post_content', $user_cpt ) ) {
+			$decoded_data = json_decode( $user_cpt['post_content'], true );
+
+			$json_decoding_error = json_last_error();
+			if ( JSON_ERROR_NONE !== $json_decoding_error ) {
+				trigger_error( 'Error when decoding a theme.json schema for user data. ' . json_last_error_msg() );
+				return new WP_Theme_JSON( $config, 'user' );
+			}
+
+			// Very important to verify if the flag isGlobalStylesUserThemeJSON is true.
+			// If is not true the content was not escaped and is not safe.
+			if (
+				is_array( $decoded_data ) &&
+				isset( $decoded_data['isGlobalStylesUserThemeJSON'] ) &&
+				$decoded_data['isGlobalStylesUserThemeJSON']
+			) {
+				unset( $decoded_data['isGlobalStylesUserThemeJSON'] );
+				$config = $decoded_data;
+			}
+		}
+		self::$user = new WP_Theme_JSON( $config, 'user' );
+
+		return self::$user;
+	}
+
+	/**
+	 * There are three sources of data (origins) for a site:
+	 * core, theme, and user. The user's has higher priority
+	 * than the theme's, and the theme's higher than core's.
+	 *
+	 * Unlike the getters {@link get_core_data},
+	 * {@link get_theme_data}, and {@link get_user_data},
+	 * this method returns data after it has been merged
+	 * with the previous origins. This means that if the same piece of data
+	 * is declared in different origins (user, theme, and core),
+	 * the last origin overrides the previous.
+	 *
+	 * For example, if the user has set a background color
+	 * for the paragraph block, and the theme has done it as well,
+	 * the user preference wins.
 	 *
 	 * @since 5.8.0
+	 * @since 5.9.0 Add user data and change the arguments.
 	 *
-	 * @param array $settings Optional. Existing block editor settings. Default empty array.
+	 * @param string $origin To what level should we merge data.
+	 *                       Valid values are 'theme' or 'user'.
+	 *                       Default is 'user'.
 	 * @return WP_Theme_JSON
 	 */
-	public static function get_merged_data( $settings = array() ) {
-		$theme_support_data = WP_Theme_JSON::get_from_editor_settings( $settings );
-
+	public static function get_merged_data( $origin = 'user' ) {
```

Comment:
Changing the method signature could be a breaking change for backwards-compatibility. However, there is a proper value and data type check `===`. Should be okay.

Reply:
Does it make any difference that this class is marked as `@private`? Hence, it's expected that it's not used by anyone.

Reply:
No not really. private is just a suggestion in WordPress, the whole code is public. 

You do either this 

```
public static function get_merged_data( $deprecated = array(), $origin = 'user' ) {
              if ( ! empty( $deprecated ) ) {
		 _deprecated_argument( __FUNCTION__, '5.9' );
	      }

```

or
```

public static function get_merged_data( $origin = 'user' ) {
              if ( is_array( $origin ) ) {
		 doing_it_wrong( 'Do not pass array here', '5.9' );
	      }
```

Reply:
>Does it make any difference that this class is marked as @private? Hence, it's expected that it's not used by anyone.

Yes, it should make a difference.

Here's my understanding of `@access private`: 

Marking a global function or class as `@access private` means: it's only used in and by Core. It's internal functionality. There's no binding contract for or guarantee of backwards-compatibility externally.

Reply:
[From the Core handbook](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#1-functions-class-methods):

>@access: Only used for core-only functions or classes implementing “private” core APIs. If the element is private it will be output with a message stating its intention for internal use.